feat: Add Firewood implementation

alarso16 opened the pull request:
## Why this should be merged

This is the first step in integrating Firewood into libevm for shared use in coreth and subnet-evm. This PR includes the `triedb.DBOverride` implementation, as well as the `state.Trie` implementations necessary to inject into `state.Database`. This is essentially a re-write of the current implementation in coreth, so feel free to suggest any structural changes. Let's get it right this time.

Note this cannot be merged until the next Firewood release. The automatic proposal cleanup and some bug fixes are necessary.

The most notable difference from coreth is a performance difference (see ava-labs/coreth#1238)

## How this works

All the logic is essentially the same as before, with the exception of proposal creation and cleanup. Proposals are now created at hash time, and then rather than being dropped, are passed to the triedb for potential commit.

## How this was tested

Sorry, there's no unit tests. I'm thinking PR 2 will be creating the `state.Database` injections and fixing the fuzz test in coreth. However, you can test this by checking out `alarso16/firewood-libevm` in coreth, using some `go mod edit` statements, and building firewood locally. It passes all coreth tests.

Alternatively, there's a DONOTMERGE PR in coreth running UT: ava-labs/coreth#1399

alarso16 committed:
feat: Firewood as triedb

powerslider commented on libevm/triedb/firewood/account_trie.go:
```diff
@@ -0,0 +1,299 @@
+	parentRoot   common.Hash
+	root         common.Hash
+	reader       database.Reader
+	dirtyKeys    map[string][]byte // Store dirty changes
```
You could use `sync.Map` as a direct replacement or add a `sync.RWMutex` to guard access to the map.

Also consider if sharing `AccountTrie` instances across goroutines is necessary, or if each goroutine should have its own copy.

alarso16 committed:
build: Add Firewood dependency

alarso16 committed:
fix: bad err check, lint

ARR4N commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+// WARNING: If anything from go-ethereum uses this,
+// it must be overwritten to use something like:
+// `_, ok := db.(*Database); if !ok { return "" }`
+// to recognize the Firewood database.
```
This can't be left to a comment here because it won't ever be done. We need to find a more robust way to enforce it, perhaps in the `triedb` package, or by returning a different scheme entirely.

ARR4N commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+	// All remaining proposals must be dereferenced.
+	db.possibleProposals = nil
+	db.proposalMap = nil
+	db.proposalTree = nil
```
Should we not just explicitly drop them? The finalizer semantics were intended as a fail-safe just in case we "lose" one in the `libevm` plumbing, but in this case it's all under our control.

ARR4N commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+
+func (db *Database) Update(root, parent common.Hash, block uint64, _ *trienode.MergedNodeSet, _ *triestate.Set, opts ...stateconf.TrieDBUpdateOption) error {
+	// We require block hashes to be provided for all blocks in production.
+	// However, many tests cannot reasonably provide a block hash for genesis, so we allow it to be omitted.
```
Production code shouldn't be hobbled to allow for improper tests. Why can't a genesis block's hash be provided?

ARR4N committed:
refactor: `firewood.Database.Update()` and `Commit()` readability (#242)

## Why this should be merged

Simplify existing implementation.

## How this works

1. Abstract code to reduce nesting
2. Restructure `Database` contents to logically group similar components
3. Embed `blockInfo` inside `proposal` as all fields effectively belong
to the `proposal` anyway

## How this was tested

N/A as base branch is still to be tested.

alarso16 commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+	logPath := filepath.Join(config.ChainDir, logFileName)
+	if err := ffi.StartLogs(&ffi.LogConfig{Path: logPath}); err != nil {
+		// This shouldn't be a fatal error, as this can only be called once per thread.
+		// Specifically, this will return an error in unit tests.
```
I totally agree, it's not great. The thing is that we can only initialize on set of logs per process right now, and the entire test package will fail after the first database open. I don't know how else we can address this. Really, the logs should be tied to a specific database instance, but this is an open ticket in firewood

alarso16 commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+		return nil, err
+	}
+
+	// Start the logs prior to opening the database
```
Yeah this is probably unnecessary. It's very clear that what and why, so I'll delete it

alarso16 commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+	return 0, 0
+}
+
+// This isn't called anywhere in coreth
```
Oops, yeah that's outdated. From triedb docs:
```
// Reference adds a new reference from a parent node to a child node. This function
// is used to add reference between internal trie node and external node(e.g. storage
// trie root), all internal trie nodes are referenced together by database itself.
```
It's only used for HashDB anyway. But yes, it's only exported as a keep-alive signal

alarso16 committed:
fix: address coding feedback

alarso16 commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+
+// Initialized checks whether a non-empty genesis block has been written.
+func (db *Database) Initialized(common.Hash) bool {
+	rootBytes, err := db.Firewood.Root()
```
Working on it! See ava-labs/firewood#1449

alarso16 committed:
fix: update go generate for new proto version

alarso16 committed:
fix: comment formatting

alarso16 committed:
feat: Drop immediately

alarso16 committed:
chore: update firewood

alarso16 committed:
Add recovery of proposalTree state

alarso16 committed:
add archive support

alarso16 commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+// This isn't called anywhere in coreth
+func (*Database) Reference(common.Hash, common.Hash) {}
+
+// Dereference drops a proposal from the database.
```
I updated these comments, so hopefully this makes more sense

alarso16 commented on libevm/triedb/firewood/firewood.go:
```diff
@@ -0,0 +1,518 @@
+			info: &blockInfo{
+				root: common.Hash(currentRoot),
+				hashes: map[common.Hash]struct{}{
+					{}: {}, // genesis block has empty hash
```
Fixed
